nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java

```diff
@@ -0,0 +1,468 @@
+import org.codehaus.groovy.runtime.ResourceGroovyMethods;
+import org.codehaus.groovy.runtime.StackTraceUtils;
+
+import org.apache.nifi.processors.groovyx.sql.OSql;
+import org.apache.nifi.processors.groovyx.util.Files;
+import org.apache.nifi.processors.groovyx.util.Validators;
+import org.apache.nifi.processors.groovyx.flow.GroovyProcessSessionWrap;
+
+import groovy.lang.GroovyShell;
+import groovy.lang.Script;
+
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.components.ValidationContext;
+
+@EventDriven
+@Tags({"script", "groovy", "groovyx"})
+@CapabilityDescription(
+        "Experimental Extended Groovy script processor. The script is responsible for "
+        + "handling the incoming flow file (transfer to SUCCESS or remove, e.g.) as well as any flow files created by "
+        + "the script. If the handling is incomplete or incorrect, the session will be rolled back.")
+@Restricted("Provides operator the ability to execute arbitrary code assuming all permissions that NiFi has.")
+@SeeAlso({})
+@DynamicProperty(name = "A script engine property to update",
+        value = "The value to set it to",
+        supportsExpressionLanguage = true,
+        description = "Updates a script engine property specified by the Dynamic Property's key with the value "
+                + "specified by the Dynamic Property's value. Use `CTL.` to access any controller services.")
+public class ExecuteGroovyScript extends AbstractProcessor {
+    public static final String GROOVY_CLASSPATH = "${groovy.classes.path}";
+
+    private static final String PRELOADS = "import org.apache.nifi.components.*;" + "import org.apache.nifi.flowfile.FlowFile;" + "import org.apache.nifi.processor.*;"
+            + "import org.apache.nifi.processor.FlowFileFilter.FlowFileFilterResult;" + "import org.apache.nifi.processor.exception.*;" + "import org.apache.nifi.processor.io.*;"
+            + "import org.apache.nifi.processor.util.*;" + "import org.apache.nifi.processors.script.*;" + "import org.apache.nifi.logging.ComponentLog;";
+
+    public static final PropertyDescriptor SCRIPT_FILE = new PropertyDescriptor.Builder().name("Script File").required(false)
+            .description("Path to script file to execute. Only one of Script File or Script Body may be used").addValidator(Validators.createFileExistsAndReadableValidator())
+            .expressionLanguageSupported(true).build();
+
+    public static final PropertyDescriptor SCRIPT_BODY = new PropertyDescriptor.Builder().name("Script Body").required(false)
+            .description("Body of script to execute. Only one of Script File or Script Body may be used").addValidator(StandardValidators.NON_EMPTY_VALIDATOR).expressionLanguageSupported(false)
+            .build();
+
+    public static String[] VALID_BOOLEANS = {"true", "false"};
+    public static final PropertyDescriptor REQUIRE_FLOW = new PropertyDescriptor.Builder().name("Requires flow file")

```

Comment:
with `require flow file = false` you can do any session.get(...) in code
i agree it's not a big win to use `require flow file = true`. 
i thought to add more values like `batch` and use `batch_size` custom property.

a short entry about `require flow file` and `fail strategy`:
we were migrating an 10years old j2ee project to nifi.  100k+ producers and 100k consumers of the files with different formats (xml,csv,json,pdf,tiff,...). you can imagine a huge amount of specific code, libraries, ... to cover all specific things we used groovy script. and in 95% we used just `session.get(0)` (require flow file)

and in 80%  `rollback` strategy  on exception is ok (btw it's a standard behavior of existing execute script processor). without try-catch the code just simpler. in case when using `rollback` you can do any custom try-catch in your code. and we used custom try-catch in 1% of cases.

transfer to `failure` in code usually becomes a nightmare in script especially if you want to transfer the original file - and in our case it was always original. 

so, both features are not removing customization ability. 

about the `require flowfile` - i agree that it could be removed.

but about `fail strategy` - i believe it's very useful. maybe the items in select should be renamed and more clarified... but i'd suggest to keep it.